bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/ConfiguredDeadlineGeneratorFactory.java

```diff
@@ -0,0 +1,166 @@
+   * CallOptions each time this method is called.
+   */
+  private static class DeadlineGeneratorImpl implements DeadlineGenerator {
+    private final boolean useTimeout;
+    private final int requestTimeoutMs;
+    private final Optional<Integer> requestAttemptTimeoutMs;
+    private final Optional<Long> nonTimeoutFallbackDeadlineMs;
+    private final Optional<Deadline> operationDeadline;
+
+    DeadlineGeneratorImpl(
+        boolean useTimeout,
+        int requestTimeoutMs,
+        Optional<Integer> requestAttemptTimeoutMs,
+        Optional<Long> nonTimeoutFallbackDeadlineMs) {
+      this.useTimeout = useTimeout;
+      this.requestTimeoutMs = requestTimeoutMs;
+      this.requestAttemptTimeoutMs = requestAttemptTimeoutMs;
+      this.nonTimeoutFallbackDeadlineMs = nonTimeoutFallbackDeadlineMs;
+      this.operationDeadline = createOperationDeadline();
+    }
+
+    /**
+     * Creates a {@link Deadline} for the operation, decided in the following order:
+     *
+     * <ol>
+     *   <li>If a user set a {@link Context} deadline (see {@link Deadline}), use that
+     *   <li>If a user configured deadlines via {@link CallOptionsConfig}, use it.
+     *   <li>Otherwise, no deadline is specified.
+     * </ol>
+     */
+    private Optional<Deadline> createOperationDeadline() {
+      Deadline contextDeadline = Context.current().getDeadline();
+      if (contextDeadline != null) {
+        return Optional.of(contextDeadline);
+      } else if (useTimeout) {
+        return Optional.of(Deadline.after(requestTimeoutMs, TimeUnit.MILLISECONDS));
+      } else {
+        return Optional.absent();
+      }
+    }
+
+    @Override
+    public Optional<Long> getOperationTimeoutMs() {
+      return operationDeadline.isPresent()
+          ? Optional.of(operationDeadline.get().timeRemaining(TimeUnit.MILLISECONDS))
+          : Optional.<Long>absent();
+    }
+
+    /**
+     * Create a {@link Deadline} that has a fail safe RPC deadline to make sure that unary
+     * operations don't hang. This will have to be overridden for streaming RPCs like read rows.
+     *
+     * @see com.google.cloud.bigtable.grpc.io.Watchdog Watchdog which handles hanging for streaming
+     *     reads (i.e. the {@link #nonTimeoutFallbackDeadlineMs} value for the streaming RPC
+     *     above)..
+     */
+    @Override
+    public Optional<Deadline> getRpcAttemptDeadline() {
+      // Check for RPC Attempt Timeouts first for backward-compatibility, so that
+      // the existing logic (based mostly on the operation timeout) is preserved.
+      if (useTimeout && requestAttemptTimeoutMs.isPresent()) {
+        Deadline attemptDeadline =
+            Deadline.after(requestAttemptTimeoutMs.get(), TimeUnit.MILLISECONDS);
+        // Ensure we don't exceed the absolute operation deadline by taking the min of the attempt
+        // and
+        // operation deadlines.
+        return Optional.of(attemptDeadline.minimum(operationDeadline.get()));
```

Comment:
I'm not sure that operationDeadline is always non-null here. If its guaranteed that if useTimeout is true then operationDeadline is set, can you add a precondition in the constructor with a comment explaining why.

And add a note here referencing it

Reply:
It is guaranteed to be non-null due to the useTimeout condition you mentioned.
I've added notes to clarify.
Not sure how strict of a precondition you want; let's talk offline.